### PR TITLE
New algorithm for volume approximation

### PR DESCRIPTION
**Pull request by TolisChal**
This is a major update for the CRAN version (1.1.0).

1) Implementation of the volume computation algorithm in <arXiv:1905.05494>. 

2) Faster implementation for random directions Hit and Run.

3) Improve R documentation.

4) Improve the names of functions and input variables.

### CODE REVIEW

File: include/convex_bodies/ballintersectconvex.h [hunk 1]
```diff
@@ -18,6 +18,7 @@ struct Ball{
     typedef typename Point::FT NT;
     typedef typename std::vector<NT>::iterator viterator;
 
+    Ball() {}
```
[the remark this answers is not in the document]
Comment (TolisChal) [reply include/convex_bodies/ballintersectconvex.h]: Done.

File: include/annealing/ratio_estimation.h [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        index = index % W + 1;
+
+        if (index == W) index = 0;
+        if (check_max_error123(val - zp * s, val + zp * s, error)) {
+            //if (print) std::cout << "final rejection ratio = " << val << " | total points = " << totCount << std::endl;
+            return val;
+        }
+
+    }
+    return val;
```
[the remark this answers is not in the document]
Comment (TolisChal) [reply include/annealing/ratio_estimation.h]: fixed with break in the loop.

File: include/annealing/hpoly_annealing.h [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+            u=1.0;
+            l=0.0;
+        }
+    }
+}
+
+
+template <class Zonotope, class HPolytope, class VT, class PointList, typename NT>
+void get_next_zonoball22(Zonotope &Z, std::vector<HPolytope> &HPolySet,
+                         HPolytope &HP2, VT Zs_max, VT Zs_min, PointList randPoints,
```
[the remark this answers is not in the document]
Comment (TolisChal) [reply include/annealing/hpoly_annealing.h]: Fixed.

File: include/annealing/hpoly_annealing.h [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+
+    rand_point_generator(Z, q, Ntot, var.walk_steps, randPoints, var);
+    HPolytope HP2 = HP;
+    if (check_converg001<Point>(HP, randPoints, p_value, up_lim, too_few, ratio, nu, alpha, false, true)) {
+        ratios.push_back(ratio);
+        return;
+    }
+    get_next_zonoball22(Z, HPolySet, HP2, Zs_max, HP.get_vec(), randPoints, ratios, p_value, up_lim, nu, alpha);
+
+    ZonoHP ZHP2;
```
[the remark this answers is not in the document]
Comment (TolisChal) [reply include/annealing/hpoly_annealing.h]: See ZonoIntersectHPoly. What do you mean by "template"?

File: include/annealing/hpoly_annealing.h [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+        if(u-l<0.00000000001) {
+            u=1.0;
+            l=0.0;
+        }
+    }
+}
+
+
+template <class Zonotope, class HPolytope, class VT, class PointList, typename NT>
+void get_next_zonoball22(Zonotope &Z, std::vector<HPolytope> &HPolySet,
```
[the remark this answers is not in the document]
Comment (TolisChal) [reply include/annealing/hpoly_annealing.h]: Fixed.

File: include/annealing/hpoly_annealing.h [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+        q=Point(n);
+        med = (u + l) * 0.5;
+        Zmed = Zs_min + (Zs_max-Zs_min)*med;
+        HPiter.set_vec(Zmed);
+
+        randPoints.clear();
+        rand_point_generator(HPiter, q, 1200, 10+2*n, randPoints, variter);
+        too_few = false;
+
+        if(check_converg001<Point>(P, randPoints, lb, up_lim, too_few, ratio, 10, 0.2, true, false)) {
```
[the remark this answers is not in the document]
Comment (TolisChal) [reply include/annealing/hpoly_annealing.h]: Done.

File: include/annealing/hpoly_annealing.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (vissarion) [reply include/annealing/hpoly_annealing.h]: I mean you can use `ZonoIntersectHPoly` type for `ZHP2` and delete the `ZonoHP` template. Unless if you think that `ZonoIntersectHPoly` can be implemented in a different way.

File: include/annealing/ratio_estimation.h [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+        m = (m - last_W[index] / NT(W)) + val / NT(W);
+        sum_sq = (sum_sq - last_W[index] * last_W[index]) + val * val;
+        sum = (sum - last_W[index]) + val;
+        s = std::sqrt((sum_sq + NT(W) * m * m - 2.0 * m * sum) / NT(W));
+        last_W[index] = val;
+        index = index % W + 1;
+
+        if (index == W) index = 0;
+        if (check_max_error123(val - zp * s, val + zp * s, error)) {
```
[the remark this answers is not in the document]
Comment (TolisChal) [reply include/annealing/ratio_estimation.h]: done

File: include/annealing/ratio_estimation.h [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+
+template <class RNGType, class Point, class PolyBall1, class PolyBall2, typename NT, class Parameters>
+NT esti_ratio_interval(PolyBall1 &Pb1, PolyBall2 Pb2, NT ratio, NT error, int W, int Ntot, NT prob,
+                       Parameters &var, bool isball = false, NT radius = 0.0) {
+
+    int n = var.n, index = 0;
+    bool print = var.verbose;
+    std::vector<NT> last_W(W,0), lamdas(Pb1.num_of_hyperplanes(),0), Av(Pb1.num_of_hyperplanes(),0);
+    NT val, countIn = ratio*NT(Ntot), totCount = NT(Ntot), sum_sq=0.0, sum=0.0, lambda;
```
[the remark this answers is not in the document]
Comment (TolisChal) [reply include/annealing/ratio_estimation.h]: done

File: include/annealing/ball_annealing.h [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+    int m = randPoints.size()/nu;
+    NT T, rs, alpha_check = 0.01, countsIn = 0.0;
+
+    bool print = true;
+
+    int i = 1;
+    for(typename std::list<Point>::iterator pit=randPoints.begin(); pit!=randPoints.end(); ++pit, i++){
+
+        if (P.is_in(*pit)==-1) countsIn += 1.0;
+        if (i % m == 0) {
```
[the remark this answers is not in the document]
Comment (TolisChal) [reply include/annealing/ball_annealing.h]: done.

**Pull request merged**